addons/stock/models/product.py

```diff
@@ -886,6 +886,23 @@ def _onchange_type(self):
             }
         return res
 
+    @api.onchange('uom_id')
+    def _onchange_uom_id(self):
+        moves = self.env['stock.move'].sudo().search_count(
+            [('product_id', 'in', self.with_context(active_test=False).product_variant_ids.ids)], limit=1
+        )
+        if moves:
+            return {
+                'warning': {
+                    'title': _('Warning!'),
+                    'message': _(
+                        'This product has been used in at least one inventory movement. '
+                        'It is not advised to change the Unit of Measure since it can lead to inconsistencies. '
+                        'The existing moves will not be recalculated with the new unit of measure.'
+                    )
+                }
+            }
```

Comment:
If we're relaxing the changing of UoMs for products with already done moves, then maybe you should also return a warning here too:

https://github.com/odoo/odoo/blob/4bda75cc8cba48ce0ce0351cbf27aad1f00f2b10/addons/stock/models/product.py#L1120-L1123

A similar warning should be included for when users change the reference UoM as well